src/x86_64/flush.c

```diff
@@ -2,54 +2,226 @@
 #include <page.h>
 #include <apic.h>
 
-// in order to keep SMP_TEST conditionalized
+#define FLUSH_THRESHOLD 32
+#define MAX_FLUSH_ENTRIES 1024
+#define COMP_QUEUE_SIZE (MAX_FLUSH_ENTRIES*2)
+#define ENTRIES_SERVICE_THRESHOLD (MAX_FLUSH_ENTRIES/2)
+
 static boolean initialized = false;
-static queue flush_queue;
 static int flush_ipi;
 static heap flush_heap;
+static volatile word inval_gen;
+static queue free_flush_entries;
+static struct list entries;
+static int entries_count;
+static volatile boolean service_scheduled;
+static thunk flush_service;
+static queue flush_completion_queue;
+static struct spinlock flush_lock;
+
+static void queue_flush_service();
 
-typedef struct flush_entry {
-    struct refcount r;
-    u64 page; // or INVALID_ADDRESS for a full flush
-} *flush_entry;
+declare_closure_struct(1, 0, void, flush_complete,
+    flush_entry, f);
+
+struct flush_entry {
+    struct list l;
+    u64 gen;
+    u64 cpu_mask;
+    struct refcount ref;
+    boolean flush;
+    u64 pages[FLUSH_THRESHOLD];
+    int npages;
+    thunk completion;
+    closure_struct(flush_complete, finish);
+};
 
 static void invalidate (u64 page)
 {
-    asm volatile("invlpg (%0)" :: "r" (page) : "memory");            
+    asm volatile("invlpg (%0)" :: "r" (page) : "memory");
 }
 
-closure_function(0, 0, void, flush_handler)
+define_closure_function(1, 0, void, flush_complete, flush_entry, f)
+{
+    flush_entry f = bound(f);
+    assert(f->cpu_mask == 0);
+    queue_flush_service();
+}
+
+/* must be called with interrupts off */
+static void _flush_handler(void)
 {
-    flush_entry f = queue_peek(flush_queue);
-    if (f->page == INVALID_PHYSICAL) {
+    cpuinfo ci = current_cpu();
+    /* Each generation has at least one page, so if the gen difference is
+     * greater than FLUSH_THRESHOLD, just do a full tlb flush */
+    boolean full_flush = inval_gen - ci->inval_gen > FLUSH_THRESHOLD;
+
+    spin_rlock(&flush_lock);
+    while (ci->inval_gen != inval_gen) {
+        word oldgen = ci->inval_gen;
+        ci->inval_gen = inval_gen;
+        list_foreach(&entries, l) {
+            flush_entry f = struct_from_list(l, flush_entry, l);
+            if (f->gen <= oldgen)
```

Comment:
I like the idea because I have seen one processor get hundreds of entries behind all the others, so being able to start mid-list would eliminate unnecessary traversal. However, when thinking through this, since the cpu would be tracking a list pointer that will eventually be removed when the list is cleaned up, it requires the cleaner to check every cpu to see if it has that pointer, and do this for every flush entry being removed. That sounds expensive enough that I'm not sure it will actually benefit, so I'm punting this for now. I think it's still worth investigating if we think of a good way to do it.